R/aes.r

```diff
@@ -334,3 +334,51 @@ mapped_aesthetics <- function(x) {
   is_null <- vapply(x, is.null, logical(1))
   names(x)[!is_null]
 }
+
+
+#' Check a mapping for discouraged usage
+#'
+#' Checks that `$` and `[[` are not used when the target *is* the data
+#'
+#' @param mapping A mapping created with [aes()]
+#' @param data The data to be mapped from
+#'
+#' @noRd
+check_aes_extract_usage <- function(mapping, data) {
+  lapply(mapping, check_aes_extract_usage_quo, data)
+}
+
+check_aes_extract_usage_quo <- function(quosure, data) {
+  check_aes_extract_usage_expr(get_expr(quosure), data, get_env(quosure))
+}
+
+check_aes_extract_usage_expr <- function(x, data, env = emptyenv()) {
+  if (is_call(x, "[[") || is_call(x, "$")) {
+    if (extract_target_is_data(x, data, env)) {
```

Comment:
Only check here if `x[[2]]` is a symbol — I don't think we need to handle cases like `foo(1,2,3)$x`